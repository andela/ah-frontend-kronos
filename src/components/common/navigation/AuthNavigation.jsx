/* eslint-disable jsx-a11y/alt-text */
/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { Link } from 'react-router-dom';
import '../../../assets/css/AuthNavigation.css';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import history from '../../../utils/history';
import { fetchProfile } from '../../../actions/profile/ProfileAction';

export class Navigation extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      image: null,
    };
  }

  componentDidMount() {
    const { fetchProfile } = this.props;
    fetchProfile(sessionStorage.getItem('username'));
  }

  componentWillReceiveProps(nextProps) {
    const { profile } = nextProps;
    this.setState({
      image: profile.profile.image,
    });
  }

  handleClick = (event) => {
    event.preventDefault();
    sessionStorage.removeItem('token');
    sessionStorage.setItem('isLoggedIn', false);
    sessionStorage.setItem('logged_in', false);
    sessionStorage.removeItem('username');
    this.setState({
      image: null,
    });
    history.push('/login');
  };

  render() {
    const { image } = this.state;
    const loggedIn = sessionStorage.getItem('isLoggedIn');
    return (
      <nav className="navbar navbar-light navbar-expand-md navbar-fixed-top navigation-clean-button">
        <div className="container">
          <Link to="/" className="navbar-brand">
            <span>Authors-haven</span>
          </Link>
          <button
            type="button"
            data-toggle="collapse"
            data-target="#navcol-1"
            className="navbar-toggler"
          >
            <span className="sr-only">Toggle navigation</span>
            <span className="navbar-toggler-icon" />
          </button>
          <div className="collapse navbar-collapse" id="navcol-1">
            <div className="ml-auto navbar-text actions">
              {
                (loggedIn === 'true')
                  ? (
                    <React.Fragment>
                      <Link to="/create-article" className="btn btn-light action-buttons">
                        Create Article
                      </Link>
                      <Link to="/articles" className="btn btn-light action-buttons">View Articles</Link>
                      <li className="nav-item dropdown show">
                        <a
                          className="dropdown-toggle nav-link"
                          id="navbarDropdownMenuLink-333"
                          data-toggle="dropdown"
                          aria-expanded="false"
                        >
                          {!image || (image === null) ? (<img className="rounded-circle" src="https://vignette.wikia.nocookie.net/caramella-girls/images/9/99/Blankpfp.png/revision/latest?cb=20190122015011" width="50hv" height="50hv" />)
                            : (<img className="rounded-circle" src={image} width="50hv" height="50hv" />)}
                          {' '}
                          {sessionStorage.getItem('username')}
                        </a>
                        <div className="dropdown-menu dropdown-default" aria-labelledby="navbarDropdownMenuLink-333">
                          <a className="dropdown-item" href="/viewProfile">View Profile</a>
                          <a className="dropdown-item" href="/editProfile">Edit Profile</a>
                          <a className="dropdown-item" onClick={this.handleClick} href="/login">Logout</a>
                        </div>
                      </li>
                    </React.Fragment>
                  )
                  : (
                    <React.Fragment>
                      <Link to="/signup" className="btn btn-light action-buttons">Sign Up</Link>
                      <Link to="/login" className="btn btn-light action-buttons">Log In</Link>
                    </React.Fragment>
                  )
              }
            </div>
          </div>
        </div>
      </nav>
    );
  }
}

Navigation.defaultProps = {
  fetchProfile: () => {},
  profile: {},
};

Navigation.propTypes = {
  fetchProfile: PropTypes.func,
  profile: PropTypes.shape({}),
};

export const mapStateToProps = state => ({
  profile: state.profile,
  loggedIn: state.loginReducer.isUserLoggedIn
    ? state.loginReducer.logged_in
    : state.socialLoginReducer.logged_in,
});

export default connect(
  mapStateToProps,
  { fetchProfile },
)(Navigation);
