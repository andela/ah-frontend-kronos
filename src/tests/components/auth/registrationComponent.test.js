import React from 'react';
import { shallow } from 'enzyme';
import Input from '../../../components/common/Input';
import Button from '../../../components/common/Button';
import { SignUpForm, mapStateToProps } from '../../../components/auth/registration/SignUpForm';
import Loading from '../../../components/common/Loading';
import { Navigation } from '../../../components/common/navigation/AuthNavigation';

function renderInputField(args) {
  const defaultProps = {
    name: '',
    type: '',
    value: '',
    handleChange: () => { },
    placeholder: '',
    onBlur: () => { },
    className: '',
    fieldError: '',
  };

  const props = { ...defaultProps, ...args };
  return shallow(<Input {...props} />);
}

function renderButtonField(args) {
  const defaultProps = {
    className: '',
    children: '',
  };
  const props = { ...defaultProps, ...args };
  return shallow(<Button {...props} />);
}

function renderSignUpForm(args) {
  const defaultProps = {
    actions: { registerUser: jest.fn() },
    user: {},
  };
  const props = { ...defaultProps, ...args };
  return shallow(<SignUpForm {...props} />);
}

function renderLoading() {
  return shallow(<Loading />);
}

function renderNavigation(args) {
  const defaultProps = {
    nextProps: {
      loggedIn: false,
    },
  };
  const props = { ...defaultProps, ...args };
  return shallow(<Navigation {...props} />);
}
it('should render Navigation', () => {
  const wrapper = renderNavigation();
  expect(wrapper).toMatchSnapshot();
});

describe('registration page', () => {
  const wrapper = renderSignUpForm();
  const wrapperInst = wrapper.instance();

  it('should render Navigation', () => {
    const wrapper = renderNavigation();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Loading in form', () => {
    const loader = renderLoading();
    expect(loader).toMatchSnapshot();
  });


  it('should render input field in form', () => {
    const inputfields = renderInputField();
    expect(inputfields.find('input').length).toBe(1);
  });

  it('should render button field in form', () => {
    const button = renderButtonField();
    expect(button.find('button').length).toBe(1);
  });

  it('form contains three input fields and one button', () => {
    const numOfInputs = wrapper.find('Input').length;
    const numOfButtons = wrapper.find('Button').length;
    expect(numOfInputs).toEqual(4);
    expect(numOfButtons).toEqual(1);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle input change', () => {
    const event = {
      target: {
        name: 'username',
        value: 'fred',
      },
    };
    wrapperInst.handleChange(event);
    expect(wrapperInst.state.username).toBe('fred');
  });

  it('should handle submit and clear form after', () => {
    wrapper.setState({
      username: 'fred',
      email: 'email@email.com',
      password: 'password',
      confirmpassword: 'confirmpassword',
    });

    expect(wrapperInst.state.username).toBe('fred');
    expect(wrapperInst.state.email).toBe('email@email.com');
    expect(wrapperInst.state.password).toBe('password');
    expect(wrapperInst.state.confirmpassword).toBe('confirmpassword');

    const event = {
      target: {
        type: 'submit',
      },
      preventDefault: jest.fn(),
    };
    wrapperInst.submitForm(event);
    expect(wrapperInst.props.actions.registerUser).toBeCalled();
    expect(wrapperInst.state.username).toBe('');
    expect(wrapperInst.state.email).toBe('');
    expect(wrapperInst.state.password).toBe('');
    expect(wrapperInst.state.confirmpassword).toBe('');
  });

  it('should map tostate to props', () => {
    const mockedState = {
      user: {
        registering: true,
      },
    };
    const state = mapStateToProps(mockedState);
    expect(state.user.registering).toBe(true);
  });
});
