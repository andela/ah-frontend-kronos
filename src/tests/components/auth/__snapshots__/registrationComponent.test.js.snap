// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration page should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`registration page should render Loading in form 1`] = `ShallowWrapper {}`;

exports[`registration page should render Navigation 1`] = `ShallowWrapper {}`;

exports[`should render Navigation 1`] = `ShallowWrapper {}`;
